pkg/graveler/retention/garbage_collection_manager.go

```diff
@@ -108,15 +112,16 @@ func (m *GarbageCollectionManager) GetRunExpiredCommits(ctx context.Context, sto
 }
 
 func (m *GarbageCollectionManager) SaveGarbageCollectionCommits(ctx context.Context, storageNamespace graveler.StorageNamespace, repositoryID graveler.RepositoryID, rules *graveler.GarbageCollectionRules, previouslyExpiredCommits []graveler.CommitID) (string, error) {
-	branchIterator, err := m.refManager.ListBranches(ctx, repositoryID)
-	if err != nil {
-		return "", fmt.Errorf("list repository branches: %w", err)
-	}
 	commitGetter := &RepositoryCommitGetter{
 		refManager:   m.refManager,
 		repositoryID: repositoryID,
 	}
-	gcCommits, err := GetGarbageCollectionCommits(ctx, branchIterator, commitGetter, rules, previouslyExpiredCommits)
+	branchIterator := ref.NewBranchIterator(ctx, m.db, repositoryID, 1000, ref.WithOrderBy("commit_id"))
+	// get all commits that are not the first parent of any commit:
+	mainAncestryLeaves := "NOT EXISTS (SELECT * FROM graveler_commits c2 WHERE c2.repository_id=graveler_commits.repository_id AND c2.parents[1]=graveler_commits.id"
```

Comment:
AFAICS this codifies that a commit can have only 2 parents.  While this is true in the current implementation, we explicitly try _not_ to make this assumption anywhere.  It may have surprising results when we *do* allow multi-merges.  Given that we are using SQL, can we avoid this?

Reply:
This code does not make assumptions about about the number of parents - the first parent is always the relevant one.

However, added handling of the parent-switch bug we used to have